tests/cpp/test_matmul_scheduler.cpp

```diff
@@ -2248,6 +2248,107 @@ TEST_F(MatmulSchedulerTest, StridedBatchEpilogueSingleBias) {
   }
 }
 
+// Test matmul with sizes that are not divisible by 8 and with misaligned inputs
+TEST_F(MatmulSchedulerTest, MisalignedVectorization) {
+  // TODO: parametrized test instead of nested loops (still use a loop over
+  // sizes and re-use FusionExecutorCache)
+  for (auto layout : kAllSupportedMmaLayout) {
+    for (bool downcast_output : {false, true}) {
+      // TODO: maybe use bias
+      auto fusion = std::make_unique<Fusion>();
+      FusionGuard fg(fusion.get());
+
+      auto tv0 = makeContigTensor(2, DataType::Half);
+      auto tv1 = makeContigTensor(2, DataType::Half);
+      fusion->addInput(tv0);
+      fusion->addInput(tv1);
+
+      tv0 = canonicalizeInputToBMNK(tv0, layout, MmaOperand::A);
+      tv1 = canonicalizeInputToBMNK(tv1, layout, MmaOperand::B);
+      auto tv2 = fusedMultiplySum(tv0, tv1, {-1});
+
+      if (downcast_output) {
+        tv2 = castOp(DataType::Half, tv2);
+      }
+
+      fusion->addOutput(tv2);
+
+      NVF_CHECK(
+          1 == ir_utils::getOpsOfType<MmaOp>(fusion.get()).size(),
+          "matmul fusion must have at least one MmaOp");
+      NVF_CHECK(
+          ir_utils::getOpsOfType<MmaOp>(fusion.get())
+              .front()
+              ->layout()
+              .has_value(),
+          "input layout has not be set for MmaOp");
+      NVF_CHECK(
+          MmaLayout::TN ==
+              ir_utils::getOpsOfType<MmaOp>(fusion.get())
+                  .front()
+                  ->layout()
+                  .value(),
+          "the MmaOp layout of Ampere MMA must be always TN");
+
+      const auto fusion_layout = mma_utils::getMmaLayout(fusion.get());
+      NVF_CHECK(
+          fusion_layout.isValid(),
+          "failed to get decide matmul layout through fusion definition");
+      NVF_CHECK(
+          fusion_layout.getData() == layout,
+          "mismatch between test layout (",
+          toString(layout),
+          ") and layout inferred from fusion definition (",
+          toString(fusion_layout.getData()),
+          ")");
+
+      FusionExecutorCache executor_cache(std::move(fusion));
+
+      for (const auto& [M, N, K, alignA, alignB, alignBias] :
+           std::vector<std::tuple<int, int, int, int, int, int>>{
+               {504, 136, 248, 16, 16, 16}, // all fully vectorizable in all
+                                            // layouts
+               {504, 136, 249, 16, 16, 16}, // odd K, operands not vectorizable
+                                            // in TN. output fully vectorizable
+               {504, 137, 248, 16, 16, 16}, // A fully vectorizable, B fully
```

Comment:
Most of the tests are passing but this one is failing in NT layout, with `downcast_output=false`. It makes an unaligned read of A somehow.
```
===== Matmul Parameters ========                                                                                           
                                                                                                                           
MMA macro: Ampere_16_16_16                                                                                                 
DoubleBufferOptions:                                                                                                       
  double_buffer_smem_write: true                                                                                           
  double_buffer_smem_read: true                                                                                            
  smem_double_buffer_stage: 2                                                                                              
SupportedVectorization:                                                                                                    
  a: 8                                                                                                                     
  b: 1                                                                                                                     
  epilogue: 1                                                                                                              
MatMulTileOptions: instruction tile [16, 16, 16], warp tile [64, 64, 32], CTA tile [256, 64, 32]                           
Rotate ldmatrix out of main loop: true                                                                                     
Async global mem load: false                                                                                               
Indexing mode: int32_t                                                                                                     
Tile rastrization order: row-major                                                                                         
Grid swizzle factor: 1                                                                                                     
Use shared memory epilogue: 0                                                                                              
Promote re-use of prologue shared memory: 1                                                                                
Split-K factor: 1                                                                                                          
==================================== 
Arguments for kernel0:                                                                                                     
Inputs:                                                                                                                    
  Tensor(sizes=[248, 504], stride=[504, 1], dtype=c10::Half, device=cuda:0, data_ptr=0x7eb2f7a00000)                       
  Tensor(sizes=[248, 137], stride=[137, 1], dtype=c10::Half, device=cuda:0, data_ptr=0x7eb2f7a3d200)                       
Outputs:                                                                                                                   
  Float [504, 137] (strides = [137, 1], address = 0x7eb2f7a4dc00)                                                          
Intermediate global buffers:  
```
```c++
  #pragma unroll
  for(nvfuser_index_t i66 = 0; i66 < 8; ++i66) {
    arraySet<__half, 8>(&T7[((i17 + (1024 * i66)) + (8LL * (i5 ^ (i18 + (4 * (i66 % 2))))))], (__half)0);
  }
  #pragma unroll
  for(nvfuser_index_t i66 = 0; i66 < 8; ++i66) {
    if ((b54 && (i55 < (-(1024LL * i66))))) {
      loadGeneric<__half, 8>( &T7[((i17 + (1024 * i66)) + (8LL * (i5 ^ (i18 + (4 * (i66 % 2))))))],  &T0[(i23 + (i24 * i66))]);
    }
  }
```

Reply:
`T0.logical_size[1LL]` is 504, which is divisible by 8. And the `data_ptr` for T0 is 0x7eb2f7a00000. The error is
> ========= Invalid __global__ read of size 16 bytes
=========     at void <unnamed>::loadGeneric<<unnamed>::__half, (int)8>(T1 *, T1 *)+0x1730 in /opt/pytorch/nvfuser/__tmp_kernel_2.cu:2383
=========     by thread (1,0,1) in block (0,0,0)
=========     Address 0x72c4d3a00402 is misaligned
=========     and is inside the nearest allocation at 0x72c4d3a00000 of size 2097152 bytes
=========     Device Frame:<unnamed>::nvfuser_2(<unnamed>::Tensor<<unnamed>::__half, (int)2, (int)2>, <unnamed>::Tensor<<unnamed>::__half, (int)2, (int)2>, <unnamed>::Tensor<float, (int)2, (int)2>)+0x1730 in /opt/pytorch/nvfuser/__tmp_kernel_2.cu:10175

Reply:
The offending address 0x7eb2f7a013ba is aligned at 2 bytes (`bin(0x7eb2f7a013ba) == '0b11111101011001011110111101000000001001110111010'`). I'm trying to figure out how this can happen.

Reply:
Thread (0, 0, 1) in block (0, 0, 0) should be trying to access these three elements:
- `i3 = 0` : `T0.logical_size[1LL]` = 504
- `i3 = 1` : `T0.logical_size[1LL] + 4 * T0.logical_size[1LL]` = 5 * 504 = 2520
- `i3 = 2` : 9 * 504 = 4536
Yet the offending offset is at 0x7eb2f7a013ba = 0x7eb2f7a00000 + 5050 bytes. 5050 bytes = 2525 `__half` elements. so it appears there's some problem at `i3=1`?

Reply:
Also suspicious: this doesn't happen if I isolate the test. That is, if I comment out the other problem sizes here, it goes away. I can set the layout to only `NT` and preserve the error..

Reply:
OK I believe this is due to kernel re-use in `FusionExecutorCache`.

Reply:
I was just missing a comparison for the new attribute `supported_vec_size` in `MatmulParams::sameAs`.